ASoC: SOF: ipc4-mtrace: New implementation

ujfalusi opened the pull request:
Hi @bardliao ,

this is a new mtrace implementation which looks up for the core to slot mapping for the debug log and have per core debugfs file.

@RanderWang, fyi

bardliao committed:
[WIP]:ASoC: SOF: add trace support for IPC4

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>
Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
[wip] ASoC: SOF: ipc4-mtrace: Update for safe and clean use

- do not enable mtrace if not supported by firmware
- do not enable mtrace on not tgl platforms as the log message is different
  for example
- enable mtrace on init and not only after first resume
- fix mtrace disable if sof/mtrace file is closed before suspend
- use correct ipc message setup
- separate log file per slots
- complete rewrite

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
Revert "[WIP]:ASoC: SOF: add trace support for IPC4"

This reverts commit e88a7b293d0b473d2d47a411b307f245e4b3c8f2.

ujfalusi committed:
ASoC: SOF: Intel: mtl: Set MTRACE type to TGL

MeteorLake is using the same type of mtrace as TGL.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
[HACK]: ipc4-mtrace: symlink the mtrace file to mtrace_root/core0

for compatibility's sake for existing tooling.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

RanderWang reviewed (commented):
@ujfalusi  Skip APL, SKL and CNL now since they are CAVS1.5 or 1.8  and use different log memory layout compared to tgl in CAVS 2.x version. Current code can't support them.

ujfalusi commented:
> @ujfalusi Skip APL, SKL and CNL now since they are CAVS1.5 or 1.8 and use different log memory layout compared to tgl in CAVS 2.x version. Current code can't support them.

Yes, the mtrace will only initialize for TGL, but I needed some way to do this in a clean way.

ujfalusi committed:
[HACK] ipc4-mtrace: move read pointer with mtrace slot size all the time

this is wrong, but it looks like the user space wants this?

We will in essence tell user space that we have given 0x1000 of data all
the time while we never do that.
Oh well.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
fixup! slot_id is not resource id!

Blah, this is now works.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>
